doc/metric.go

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package doc
+
+// NewMetric returns a new metric document.
+func NewMetric(id []byte, tags map[string]string) Document {
```

Comment:
> Any reason to not make this a more generic NewDoc function instead?

Sure, I'll rename. 

> Perhaps the value type of the map should be []byte?

I'd like to eventually structure it so it doesn't require any allocs. So we might end up changing the signature to something like 

` func NewDoc(id []byte, tagKeys [][]byte, tagValues [][]byte) Document {... `

I'll put a TODO down for now, we can revisit once we see how inter-op with m3db works. 

> Is it worth calling out that field names must be valid UTF-8?

Hmm yeah that's a fair point. I'll put it in the same note as above. 

> Also, it just occurred to me that we'll probably want to index the ID's so perhaps we to have some reserved field names? Or maybe we just accept a map of fields and if _id is set then we use that as the ID, otherwise we generate an ID ourselves? FWIW this seems to be the approach taken by Elasticsearch

I was going the route of being more strict about this. `ID` has to be set, and it's the unique identifier for a document. The internal doc id we assign to a document won't matter to users of the API.

Reply:
All sound good to me, but perhaps we should have `NewDoc` and `NewDocWithID` with the former being a convenience wrapper for the latter where we generate the ID ourselves?

Reply:
Thought about the `_id` suggestion some more, I like it. 

Re: generating the ID, I don't think the index should do that. If people don't have an ID that fits into their natural scheme, I'd prefer they generated the `UUID` or w/e we'd end up doing internally. That way they're able to define uniqueness for their use-case.

Reply:
Yea, I just feel like for most use cases there won't in fact be a unique ID. It's easy to add such a thing in the future though if the need arises.

Reply:
done